src/client/opamCommands.ml

```diff
@@ -2391,18 +2421,28 @@ let switch =
       let switch_name = OpamSwitch.of_string switch in
       OpamSwitchCommand.switch `Lock_none gt switch_name;
       `Ok ()
-    | Some `set_compiler, packages ->
-      (try
-         let parse_namev s = match fst OpamArg.package s with
-           | `Ok (name, version_opt) -> name, version_opt
-           | `Error e -> failwith e
-         in
-         let namesv = List.map parse_namev packages in
-         OpamGlobalState.with_ `Lock_none @@ fun gt ->
+    | Some `show_invariant, [] ->
+      OpamGlobalState.with_ `Lock_none @@ fun gt ->
+      OpamSwitchState.with_ `Lock_none gt @@ fun st ->
+      OpamConsole.msg "%s\n"
+        (OpamFileTools.dep_formula_to_string st.switch_invariant);
+      `Ok ()
+    | Some `set_invariant, params ->
+      OpamGlobalState.with_ `Lock_none @@ fun gt ->
+      OpamRepositoryState.with_ `Lock_none gt @@ fun rt ->
+      (match invariant_arg rt params with
+       | exception Failure e -> `Error (false, e)
+       | invariant ->
          OpamSwitchState.with_ `Lock_write gt @@ fun st ->
-         OpamSwitchState.drop @@ OpamSwitchCommand.set_compiler st namesv;
-         `Ok ()
-       with Failure e -> `Error (false, e))
+         let st = OpamSwitchCommand.set_invariant ~force st invariant in
+         let st =
+           if no_action || OpamFormula.satisfies_depends st.installed invariant
+           then st
+           else OpamClient.install_t
+               st ~ask:true [] None ~deps_only:false ~assume_built:false
+         in
+         OpamSwitchState.drop st;
```

Comment:
No drop with a `with_`?

Reply:
It's not needed when using `with_`, but here it's a different, modified `st` that we would need to ignore anyway ; so calling `drop` is a bit better than `ignore`...